File: internal/identity/identity.go
// Package identity is the boundary that keeps this pipeline's public actions
// under one specific GitHub account.
//
// This machine's ambient git and gh configuration resolves to a different
// account: the global user.email is that account's noreply address, and the
// global credential helper serves its token to any HTTPS push. Identity here
// is therefore never inherited -- it is asserted per clone, before any commit
// or push, across four independent layers:
//
//  1. API auth   -- GH_TOKEN per process, overriding gh's stored account
//  2. Push auth  -- a per-clone credential helper, with the chain reset first
//  3. Commit id  -- per-clone user.name/user.email plus GIT_AUTHOR_*
//  4. Guards     -- shared git hooks that abort a wrong-identity push
//
// Layer 4 is not trusted alone: a repo's own tooling can rewrite
// core.hooksPath (husky does this on npm install) and silently disarm it,
// which is why AssertClone re-checks all of the above in-process.
package identity

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// CredentialKey is the git config key whose chain decides which token is
// served to a push. Getting this wrong pushes as the wrong account.
const CredentialKey = "credential.https://github.com.helper"

// Error reports that a clone is not provably operating as the OSS identity.
// It carries every violation found, not just the first: when this fires in an
// unattended run, the whole picture is what makes it debuggable.
type Error struct{ Problems []string }

func (e *Error) Error() string {
	return fmt.Sprintf("identity not asserted:\n  - %s",
		strings.Join(e.Problems, "\n  - "))
}

// Identity is the account everything public happens under, plus the other
// accounts on this machine that must never be used by mistake.
type Identity struct {
	Login           string
	Name            string
	Email           string
	KeychainAccount string
	KeychainService string
	WorkLogin       string
	WorkEmail       string
	OtherLogins     []string
	OtherEmails     []string
	Root            string
}

func (i Identity) HelperPath() string { return filepath.Join(i.Root, "bin", "gh-token-helper") }
func (i Identity) HooksPath() string  { return filepath.Join(i.Root, "githooks") }

// Load parses config/identity.env.
//
// Hand-rolled rather than sourced through a shell: this is the
// security-critical path and it must not execute anything. The same file is
// read by the git hooks, which have no interpreter beyond sh -- hence the
// KEY="value" format rather than YAML.
//
// (v1 had this parsed correctly here, but read elsewhere by grepping the file,
// which matched a comment line and produced a corrupted commit author. One
// parser, used everywhere.)
func Load(path string) (Identity, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return Identity{}, fmt.Errorf("read identity: %w", err)
	}
	vals := map[string]string{}
	for _, raw := range strings.Split(string(b), "\n") {
		line := strings.TrimSpace(raw)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		key, val, ok := strings.Cut(line, "=")
		if !ok {
			continue
		}
		vals[strings.TrimSpace(key)] = unquote(val)
	}

	root, err := filepath.Abs(filepath.Dir(filepath.Dir(path)))
	if err != nil {
		return Identity{}, err
	}
	id := Identity{
		Login:           vals["OSS_LOGIN"],
		Name:            vals["OSS_NAME"],
		Email:           vals["OSS_EMAIL"],
		KeychainAccount: vals["OSS_KEYCHAIN_ACCOUNT"],
		KeychainService: vals["OSS_KEYCHAIN_SERVICE"],
		WorkLogin:       vals["WORK_LOGIN"],
		WorkEmail:       vals["WORK_EMAIL"],
		OtherLogins:     strings.Fields(vals["OTHER_LOGINS"]),
		OtherEmails:     strings.Fields(vals["OTHER_EMAILS"]),
		Root:            root,
	}
	var missing []string
	for k, v := range map[string]string{
		"OSS_LOGIN": id.Login, "OSS_NAME": id.Name, "OSS_EMAIL": id.Email,
		"OSS_KEYCHAIN_ACCOUNT": id.KeychainAccount,
		"OSS_KEYCHAIN_SERVICE": id.KeychainService,
	} {
		if v == "" {
			missing = append(missing, k)
		}
	}
	if len(missing) > 0 {
		return Identity{}, fmt.Errorf("%s is missing %s", path, strings.Join(missing, ", "))
	}
	return id, nil
}

// unquote strips one layer of surrounding quotes and any trailing comment.
// A bare value keeps everything up to the first unquoted '#'.
func unquote(v string) string {
	v = strings.TrimSpace(v)
	if len(v) >= 2 {
		if (v[0] == '"' && strings.Contains(v[1:], `"`)) ||
			(v[0] == '\'' && strings.Contains(v[1:], `'`)) {
			q := v[0]
			if end := strings.IndexByte(v[1:], q); end >= 0 {
				return v[1 : 1+end]
			}
		}
	}
	if i := strings.IndexByte(v, '#'); i >= 0 {
		v = v[:i]
	}
	return strings.TrimSpace(v)
}

// Token reads the PAT from the macOS Keychain. Never cached, never written to
// disk, never placed in a file this process controls.
func Token(id Identity) (string, error) {
	out, err := exec.Command("security", "find-generic-password",
		"-a", id.KeychainAccount, "-s", id.KeychainService, "-w").Output()
	if err != nil {
		return "", fmt.Errorf(
			"no Keychain item %s/%s; create it with:\n"+
				"  security add-generic-password -a %s -s %s -w",
			id.KeychainService, id.KeychainAccount, id.KeychainAccount, id.KeychainService)
	}
	return strings.TrimSpace(string(out)), nil
}

// Env is the environment for subprocesses that talk to GitHub or write
// commits as us -- our own git and gh calls, and nothing else.
//
// GH_TOKEN takes precedence over gh's keyring, so this overrides the machine's
// active account per process. That is the whole reason `gh auth switch` is
// banned: it mutates global state a scheduled job cannot reason about, while
// the user switches accounts by hand in parallel.
//
// Target-repo commands must NOT get this. See SandboxEnv.
func Env(id Identity, token string) []string {
	env := os.Environ()
	out := make([]string, 0, len(env)+7)
	for _, kv := range env {
		k, _, _ := strings.Cut(kv, "=")
		switch k {
		case "GH_TOKEN", "GITHUB_TOKEN": // would race with the value we set
		default:
			out = append(out, kv)
		}
	}
	return append(out,
		"GH_TOKEN="+token,
		"GIT_AUTHOR_NAME="+id.Name,
		"GIT_AUTHOR_EMAIL="+id.Email,
		"GIT_COMMITTER_NAME="+id.Name,
		"GIT_COMMITTER_EMAIL="+id.Email,
		"GIT_TERMINAL_PROMPT=0",
	)
}

// SandboxEnv is the environment for running a TARGET repo's own commands.
//
// `go test`, `npm test` and friends execute code the other project controls.
// They have no business seeing our token, so every credential-shaped variable
// is stripped -- not just ours. In v2 these run in a container as well; this
// is the belt to that braces.
func SandboxEnv() []string {
	var out []string
	for _, kv := range os.Environ() {
		k, _, _ := strings.Cut(kv, "=")
		if looksSecret(k) {
			continue
		}
		out = append(out, kv)
	}
	return append(out, "GIT_TERMINAL_PROMPT=0")
}

func looksSecret(k string) bool {
	u := strings.ToUpper(k)
	switch u {
	case "GH_TOKEN", "GITHUB_TOKEN", "ANTHROPIC_API_KEY", "CLAUDE_CODE_OAUTH_TOKEN":
		return true
	}
	for _, suffix := range []string{"_TOKEN", "_SECRET", "_KEY", "_PASSWORD", "_CREDENTIALS"} {
		if strings.HasSuffix(u, suffix) {
			return true
		}
	}
	return strings.HasPrefix(u, "AWS_")
}

func git(clone string, args ...string) (string, error) {
	out, err := exec.Command("git", append([]string{"-C", clone}, args...)...).Output()
	return strings.TrimSpace(string(out)), err
}

// HardenClone applies layers 2-4 to a fresh clone. Must run before any commit.
func HardenClone(clone string, id Identity) error {
	steps := [][]string{
		{"config", "--local", "user.name", id.Name},
		{"config", "--local", "user.email", id.Email},
		// The empty value resets the helper chain, discarding the global
		// helper that would otherwise serve the other account's token. Order
		// matters: git reads system -> global -> local, so a local reset
		// clears everything ahead of it.
		{"config", "--local", "--replace-all", CredentialKey, ""},
		{"config", "--local", "--add", CredentialKey, id.HelperPath()},
		// One canonical hooks directory rather than per-clone copies: no
		// drift, and an untrusted repo's own hooks never execute.
		{"config", "--local", "core.hooksPath", id.HooksPath()},
	}
	for _, s := range steps {
		if _, err := git(clone, s...); err != nil {
			return fmt.Errorf("harden %s: git %s: %w", clone, strings.Join(s, " "), err)
		}
	}
	return ExcludeAgentScaffolding(clone)
}

// agentScaffolding is what an implementer may leave behind in a clone.
//
// Several target repositories ship their own `.claude/skills/...`, and an
// auto-fix run has twice produced a copy under `.agents/` -- which `git add -A`
// would then commit into a public pull request. The submit guard catches that,
// but catching is weaker than preventing.
var agentScaffolding = []string{".agents/", ".aider*", ".cursor/", ".codex/"}

// ExcludeAgentScaffolding makes agent leftovers unstageable in this clone.
//
// Written to .git/info/exclude rather than .gitignore: it is our local
// concern, and editing a repository's tracked .gitignore would appear in the
// diff we are about to ask someone to merge.
func ExcludeAgentScaffolding(clone string) error {
	path := filepath.Join(clone, ".git", "info", "exclude")
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	existing, _ := os.ReadFile(path)
	var missing []string
	for _, p := range agentScaffolding {
		if !strings.Contains(string(existing), p) {
			missing = append(missing, p)
		}
	}
	if len(missing) == 0 {
		return nil
	}
	out := string(existing)
	if out != "" && !strings.HasSuffix(out, "\n") {
		out += "\n"
	}
	out += "# added by oss-pipeline: never stage agent scaffolding\n" +
		strings.Join(missing, "\n") + "\n"
	return os.WriteFile(path, []byte(out), 0o644)
}

// EffectiveCredentialHelpers is the chain git will actually consult.
//
// An empty entry clears every helper declared before it, so only the entries
// after the last empty one are live. Reading the raw list without applying
// that rule is how a reset can look like a leak, or worse, the reverse.
func EffectiveCredentialHelpers(clone string) []string {
	out, _ := git(clone, "config", "--get-all", CredentialKey)
	if out == "" {
		return nil
	}
	vals := strings.Split(out, "\n")
	last := -1
	for i, v := range vals {
		if strings.TrimSpace(v) == "" {
			last = i
		}
	}
	return vals[last+1:]
}

// AssertClone is the pre-flight check before every commit and every push.
//
// Independent of the git hooks by design: this is what still holds when a
// repo's tooling has rewritten core.hooksPath out from under us.
func AssertClone(clone string, id Identity) error {
	var problems []string

	name, _ := git(clone, "config", "--local", "user.name")
	email, _ := git(clone, "config", "--local", "user.email")
	if email != id.Email {
		hint := ""
		for i, login := range id.OtherLogins {
			if i < len(id.OtherEmails) && email == id.OtherEmails[i] {
				which := "another of your"
				if login == id.WorkLogin {
					which = "the WORK"
				}
				hint = fmt.Sprintf(" (this is %s account %s)", which, login)
				break
			}
		}
		problems = append(problems,
			fmt.Sprintf("local user.email is %q, expected %q%s", email, id.Email, hint))
	}
	if name != id.Name {
		problems = append(problems,
			fmt.Sprintf("local user.name is %q, expected %q", name, id.Name))
	}

	helpers := EffectiveCredentialHelpers(clone)
	if len(helpers) != 1 || helpers[0] != id.HelperPath() {
		problems = append(problems, fmt.Sprintf(
			"credential helper chain is %q, expected [%q] -- another account's "+
				"token could be served to a push", helpers, id.HelperPath()))
	}

	hooks, _ := git(clone, "config", "--local", "core.hooksPath")
	if hooks != id.HooksPath() {
		problems = append(problems, fmt.Sprintf(
			"core.hooksPath is %q, expected %q -- layer 4 guards are disarmed "+
				"(did a package manager rewrite it?)", hooks, id.HooksPath()))
	}

	if len(problems) > 0 {
		return &Error{Problems: problems}
	}
	return nil
}

// IsIdentityError reports whether err came from this package's assertions.
func IsIdentityError(err error) bool {
	var e *Error
	return errors.As(err, &e)
}

File: src/oss_pipeline/identity.py
"""The identity boundary.

Everything this pipeline does in public happens under the ``vimalyad`` account.
This machine's ambient git/gh configuration resolves to the *work* account --
global ``user.email`` is that account's noreply address, and the
global credential helper is ``gh auth git-credential``, which serves the work
token to any HTTPS push. Identity here is therefore never *inherited*; it is
asserted per clone, before any commit or push.

Four independent layers (plan section 1). This module owns 1-3 and re-asserts 4,
because ``core.hooksPath`` can be clobbered by a repo's own tooling -- husky does
exactly this on ``npm install`` -- which would silently disarm the git hooks.
Never trust the hooks alone; ``assert_clone`` runs in-process before every push.
"""

from __future__ import annotations

import os
import shlex
import subprocess
from dataclasses import dataclass
from pathlib import Path

CREDENTIAL_KEY = "credential.https://github.com.helper"

ROOT = Path(__file__).resolve().parents[2]


class IdentityError(RuntimeError):
    """A clone is not provably operating as the OSS identity.

    Raised with *every* violation found, not just the first: when this fires
    during an unattended run the whole picture is what makes it debuggable.
    """


@dataclass(frozen=True)
class Identity:
    login: str
    name: str
    email: str
    keychain_account: str
    keychain_service: str
    work_login: str
    work_email: str
    other_logins: tuple[str, ...] = ()
    other_emails: tuple[str, ...] = ()
    root: Path = ROOT

    @property
    def helper_path(self) -> Path:
        return self.root / "bin" / "gh-token-helper"

    @property
    def hooks_path(self) -> Path:
        return self.root / "githooks"


def load_identity(env_file: Path | None = None) -> Identity:
    """Parse ``config/identity.env``.

    Deliberately a hand-rolled parser rather than sourcing the file through a
    shell: this is the security-critical path and it must not execute anything.
    The same file is read by the git hooks, which have no interpreter available
    beyond ``sh`` -- hence the KEY="value" format rather than YAML.
    """
    path = env_file or ROOT / "config" / "identity.env"
    values: dict[str, str] = {}
    for raw in path.read_text().splitlines():
        line = raw.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, _, val = line.partition("=")
        values[key.strip()] = shlex.split(val)[0] if val.strip() else ""

    try:
        return Identity(
            login=values["OSS_LOGIN"],
            name=values["OSS_NAME"],
            email=values["OSS_EMAIL"],
            keychain_account=values["OSS_KEYCHAIN_ACCOUNT"],
            keychain_service=values["OSS_KEYCHAIN_SERVICE"],
            work_login=values["WORK_LOGIN"],
            work_email=values["WORK_EMAIL"],
            other_logins=tuple(values.get("OTHER_LOGINS", "").split()),
            other_emails=tuple(values.get("OTHER_EMAILS", "").split()),
            root=path.resolve().parents[1],
        )
    except KeyError as exc:
        raise IdentityError(f"{path} is missing {exc.args[0]}") from exc


def token(ident: Identity | None = None) -> str:
    """Read the PAT from the macOS Keychain. Never cached, never written down."""
    ident = ident or load_identity()
    proc = subprocess.run(
        ["security", "find-generic-password",
         "-a", ident.keychain_account, "-s", ident.keychain_service, "-w"],
        capture_output=True, text=True,
    )
    if proc.returncode != 0:
        raise IdentityError(
            f"No Keychain item {ident.keychain_service}/{ident.keychain_account}. "
            f"Create it with:\n"
            f"  security add-generic-password -a {ident.keychain_account} "
            f"-s {ident.keychain_service} -w"
        )
    return proc.stdout.strip()


def pipeline_env(ident: Identity | None = None) -> dict[str, str]:
    """Environment for every subprocess that talks to GitHub or writes commits.

    ``GH_TOKEN`` takes precedence over gh's keyring, so this overrides the
    machine's active account per-process. That is the whole reason
    ``gh auth switch`` is banned: it mutates global state a scheduled job cannot
    reason about, and the user switches accounts by hand in parallel.
    """
    ident = ident or load_identity()
    env = dict(os.environ)
    env.update(
        GH_TOKEN=token(ident),
        GIT_AUTHOR_NAME=ident.name,
        GIT_AUTHOR_EMAIL=ident.email,
        GIT_COMMITTER_NAME=ident.name,
        GIT_COMMITTER_EMAIL=ident.email,
        GIT_TERMINAL_PROMPT="0",
    )
    env.pop("GITHUB_TOKEN", None)  # would otherwise race with GH_TOKEN
    return env


def _git(clone: Path, *args: str, check: bool = True) -> str:
    proc = subprocess.run(
        ["git", "-C", str(clone), *args],
        capture_output=True, text=True,
    )
    if check and proc.returncode != 0:
        raise IdentityError(
            f"git {' '.join(args)} failed in {clone}: {proc.stderr.strip()}"
        )
    return proc.stdout.strip()


def harden_clone(clone: Path, ident: Identity | None = None) -> None:
    """Apply layers 2-4 to a fresh clone. Must run before the first commit."""
    ident = ident or load_identity()

    _git(clone, "config", "--local", "user.name", ident.name)
    _git(clone, "config", "--local", "user.email", ident.email)

    # The empty value resets the helper chain, discarding the global
    # `gh auth git-credential` that would otherwise serve the work token.
    # Order matters: git reads system -> global -> local, so a local reset
    # clears everything ahead of it.
    _git(clone, "config", "--local", "--replace-all", CREDENTIAL_KEY, "")
    _git(clone, "config", "--local", "--add", CREDENTIAL_KEY, str(ident.helper_path))

    # One canonical hooks directory rather than copies per clone: no drift, and
    # it also means we do not execute an untrusted repo's own hooks.
    _git(clone, "config", "--local", "core.hooksPath", str(ident.hooks_path))

    exclude_agent_scaffolding(clone)


# Agent scaffolding the implementer may create in a clone. Several repos ship
# their own `.claude/skills/...`, and an auto-fix run has twice produced a copy
# under `.agents/` -- which `git add -A` would then commit into a public PR.
#
# The submit guard catches it, but catching is weaker than preventing: listing
# these in the clone's private exclude file means `git add -A` never stages
# them in the first place. The guard stays as the second layer.
AGENT_SCAFFOLDING = (".agents/", ".aider*", ".cursor/", ".codex/")


def exclude_agent_scaffolding(clone: Path) -> None:
    """Make agent scaffolding unstageable in this clone.

    Written to .git/info/exclude rather than .gitignore: it is our local
    concern, and modifying a repo's tracked .gitignore would show up in the
    diff we are about to ask someone to merge.
    """
    path = clone / ".git" / "info" / "exclude"
    path.parent.mkdir(parents=True, exist_ok=True)
    existing = path.read_text() if path.exists() else ""
    missing = [p for p in AGENT_SCAFFOLDING if p not in existing]
    if not missing:
        return
    header = "" if existing.endswith("\n") or not existing else "\n"
    path.write_text(existing + header
                    + "# added by oss-pipeline: never stage agent scaffolding\n"
                    + "\n".join(missing) + "\n")


def effective_credential_helpers(clone: Path) -> list[str]:
    """The helper chain git will actually consult, after reset semantics.

    An empty entry clears every helper declared before it, so only the entries
    following the last empty one are live.
    """
    out = _git(clone, "config", "--get-all", CREDENTIAL_KEY, check=False)
    values = out.split("\n") if out else []
    resets = [i for i, v in enumerate(values) if v == ""]
    return values[resets[-1] + 1:] if resets else values


def assert_clone(clone: Path, ident: Identity | None = None) -> None:
    """Pre-flight check. Called before every commit and every push.

    Independent of the git hooks by design -- this is what still holds when a
    repo's tooling has rewritten ``core.hooksPath`` out from under us.
    """
    ident = ident or load_identity()
    problems: list[str] = []

    name = _git(clone, "config", "--local", "user.name", check=False)
    email = _git(clone, "config", "--local", "user.email", check=False)
    if email != ident.email:
        hint = ""
        for login, alt in zip(ident.other_logins, ident.other_emails):
            if email == alt:
                which = "WORK" if login == ident.work_login else "another of your"
                hint = f" (this is the {which} account {login})"
                break
        problems.append(f"local user.email is {email!r}, expected {ident.email!r}{hint}")
    if name != ident.name:
        problems.append(f"local user.name is {name!r}, expected {ident.name!r}")

    helpers = effective_credential_helpers(clone)
    if helpers != [str(ident.helper_path)]:
        problems.append(
            f"credential helper chain is {helpers!r}, expected "
            f"[{str(ident.helper_path)!r}] -- the work token could be served"
        )

    hooks = _git(clone, "config", "--local", "core.hooksPath", check=False)
    if hooks != str(ident.hooks_path):
        problems.append(
            f"core.hooksPath is {hooks!r}, expected {str(ident.hooks_path)!r} "
            f"-- layer 4 guards are disarmed (did a package manager rewrite it?)"
        )

    for remote in _git(clone, "remote", check=False).split("\n"):
        if not remote:
            continue
        url = _git(clone, "remote", "get-url", "--push", remote, check=False)
        if remote == "fork" and f"/{ident.login}/" not in url and f":{ident.login}/" not in url:
            problems.append(f"remote {remote!r} push url {url!r} is not a {ident.login} fork")

    if problems:
        raise IdentityError(
            f"{clone} is not provably the OSS identity:\n"
            + "\n".join(f"  - {p}" for p in problems)
        )


def assert_token_account(ident: Identity | None = None) -> str:
    """Confirm the Keychain token really belongs to the OSS account."""
    ident = ident or load_identity()
    proc = subprocess.run(
        ["gh", "api", "user", "--jq", ".login"],
        capture_output=True, text=True, env=pipeline_env(ident),
    )
    if proc.returncode != 0:
        raise IdentityError(f"gh api user failed: {proc.stderr.strip()}")
    login = proc.stdout.strip()
    if login != ident.login:
        raise IdentityError(
            f"Keychain token authenticates as {login!r}, expected {ident.login!r}. "
            f"Refusing to continue."
        )
    return login
